### PR TITLE
CVSB-1767 - Implement contingency measure

### PR DESCRIPTION
**Pull request by acozianu**

### CODE REVIEW

File: src/services/ActivityService.ts [hunk 1]
```diff
@@ -105,4 +108,117 @@ export class ActivityService {
...
+        }
+    }
+
+    /**
+     * Order Activities desc by Start Time
+     * @param activities Array of activities
+     */
+    private returnOrderedActivities(activities: string[]): string[] {
+        const sortDateDesc = (date1: any, date2: any) => {
+            if (new Date(date1.startTime) > new Date(date2.startTime)) { return -1; }
```
Comment (gmlazu) [inline src/services/ActivityService.ts]: Comparison between instances always yields `false`.

File: serverless.yml [hunk 2]
```diff
@@ -20,6 +20,9 @@ functions:
     - http:
         path: activities/{id}/end
         method: put
+    - http:
```
Comment (gmlazu) [inline serverless.yml]: New endpoint does not have integration tests.

File: tests/integration/activities.intTest.ts [hunk 3]
```diff
@@ -68,34 +68,93 @@ describe("PUT /activities/:id/end", () => {
...
         });
     });
 
     after((done: Done) => {
         activityService.dbClient.delete(postedActivity)
             .then(() => done());
     });
+});
 
+describe("GET /activities/details", () => {
```
Comment (adpetrescu) [inline tests/integration/activities.intTest.ts]: Can you please split the integration tests in multiple files? I would suggest the split to be done by endpoints.

File: tests/unit/activities.unitTest.ts [hunk 4]
```diff
@@ -315,26 +294,116 @@ describe("endActivity", () => {
...
+                .catch((error: HTTPResponse) => {
+                    const body: any = JSON.parse(error.body);
+                    expect(body.error).to.equal(`Activity already ended`);
+                });
+        });
+    });
+
+});
+
+describe("getActivities", () => {
```
Comment (adpetrescu) [inline tests/unit/activities.unitTest.ts]: Can you please split the unit tests in multiple files?

File: src/utils/Filters.ts [hunk 5]
```diff
@@ -0,0 +1,104 @@
+import { isAfter, isBefore, isEqual } from "date-fns";
+import { HTTPResponse } from "../utils/HTTPResponse";
+
+export class ActivityFilters {
+    /**
+     * Filter activities by received parameters
+     * @param response Data received from Dynamodb
+     * @param event Event
```
[the remark this answers is not in the document]
Comment (acozianu) [reply src/utils/Filters.ts]: done

File: src/services/ActivityService.ts [hunk 6]
```diff
@@ -105,4 +110,20 @@ export class ActivityService {
             });
     }
 
+    /**
+     * Get activities from Dynamodb
+     * @param event
```
[the remark this answers is not in the document]
Comment (acozianu) [reply src/services/ActivityService.ts]: done

File: src/utils/Filters.ts [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+                if (event.queryStringParameters.fromStartTime) {
+                    filteredActivities = this.filterActivitiesByStartTime(response.Items, event.queryStringParameters.fromStartTime, true);
+                } else {
+                    throw new HTTPResponse(400, "Bad request");
+                }
+                if (event.queryStringParameters.toStartTime) {
+                    filteredActivities = this.filterActivitiesByStartTime(filteredActivities, event.queryStringParameters.toStartTime, false);
+                }
+                if (event.queryStringParameters.activityType) {
+                    filteredActivities = this.filterActivitiesByParameter(filteredActivities, event.queryStringParameters.activityType, "activityType");
```
[the remark this answers is not in the document]
Comment (acozianu) [reply src/utils/Filters.ts]: done

File: src/utils/Filters.ts [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+                    break;
+                case false:
+                    if (isBefore(element.startTime, startTime) || isEqual(element.startTime, startTime)) {
+                        activityArray.push(element);
+                    }
+            }
+        });
+        if (activityArray.length) {
+            return activityArray;
+        } else {
```
[the remark this answers is not in the document]
Comment (acozianu) [reply src/utils/Filters.ts]: done

File: src/utils/Filters.ts [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+     * @param value Value of the @param field
+     * @param field activityType | testStationPNumber | testerStaffId
+     */
+    private filterActivitiesByParameter(activities: string[], value: string, field: string) {
+        const filteredArrayOfActivities = activities.filter((element: any) => {
+            return element[field] === value;
+        });
+        if (filteredArrayOfActivities.length) {
+            return filteredArrayOfActivities;
+        } else {
```
[the remark this answers is not in the document]
Comment (acozianu) [reply src/utils/Filters.ts]: done

File: tests/unit/activities.unitTest.ts (hunk 4 above)
Comment (acozianu) [reply tests/unit/activities.unitTest.ts]: done

**Comment by acozianu** [comment]
CVSB-1767

**Pull request merged**